File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 1:
            return -1 if target != nums[0] else 0
        
        l = 0
        r = len(nums) - 1
        while l <= r:
            m = l + (r - l) // 2
            if nums[m] == target:
                return m
            # left half sorted
            if nums[l] <= nums[m]:
                if nums[l] <= target and target < nums[m]:
                    r = m -1
                else:
                    l = m + 1
            else:
                if nums[m] < target and target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        return -1

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:

    vector<int> currComb;
    vector<vector<int>> result;

    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        dfs(0, target, nums);
        return result;
    }

    void dfs(int index, int target, vector<int>& nums) {

        if (target == 0) {
            result.push_back(currComb);
            return;
        }
        if (index >= nums.size() or target < 0) {
            return;
        }
        currComb.push_back(nums[index]);
        dfs(index, target - nums[index], nums);
        currComb.pop_back();
        dfs(index + 1, target, nums);
    }
};
